### PR TITLE
RT: Disruption pass midnight

### PR DESCRIPTION
**Pull request by pbougue**
Managing disruption cancellation !

JIRA: http://jira.canaltp.fr/browse/NAVP-408

One possible "problem" with this PR is it would probably require a refactor to relocate part of the code (out of type.cpp) and possibly pool some code (on cancel vj maybe)

[Review on Reviewable](https://reviewable.io/reviews/canaltp/navitia/1291)

### CODE REVIEW

File: source/kraken/apply_disruption.cpp [hunk 1]
```diff
@@ -115,24 +116,41 @@ struct apply_impacts_visitor : public boost::static_visitor<> {
...
+        if (! production_period.contains(start_date)
+                || impact.aux_info.stop_times.empty()) {
+            continue;
         }
+        // we may impact vj's passing midnight but all we care is start date
+        auto day = (start_date - production_period.begin()).days();
+        int day_offset = impact.aux_info.stop_times.front().arrival_time / ndtu::SECONDS_PER_DAY;
+        // cannot happen as arrival_time is unsigned, but in order to advance vj,
+        // it should be possible to have first stop_time negative
+        if (impact.aux_info.stop_times.front().arrival_time < 0) {
```
Comment (pbougue) [inline source/kraken/apply_disruption.cpp]: warning here at compilation (arrival-time is unsigned): what do you prefer ?
- signed cast
- completely remove this (and we might never think of it before doing a test on forward disruption if we decide to handle them)
- other ?

Comment (TeXitoi) [inline source/kraken/apply_disruption.cpp]: I think it's checked in creation of the impact from the trip_update (or it should be). Thus when we want to support it, we'll have to change that.

Comment (pbougue) [inline source/kraken/apply_disruption.cpp]: So second solution as this code does not correctly handle midnight case:
-24:00:00 => -1 day => -2 day
... Not good!
So code removal, and we'll deal with it when it's usefull and tested.

File: source/jormungandr/tests/kirin_realtime_tests.py [hunk 2]
```diff
@@ -153,6 +153,73 @@ def test_vj_delay(self):
...
+        eq_(get_arrivals(later_response), ['20120614T080935', '20120614T180222']) # pt_walk + vj 18:01
+        eq_(get_used_vj(later_response), [[], ['vj:B:1']])
+
+
+        # no disruption yet
+        pt_response = self.query_region('vehicle_journeys/vjA?_current_datetime=20120614T1337')
+        eq_(len(pt_response['disruptions']), 0)
+
+        self.send_mock("vjA", "20120614", 'delayed',
+           [("stop_point:stopB", str_to_time_stamp("20120615T070224"), str_to_time_stamp("20120615T070224")),
```
Comment (antoine-de) [inline source/jormungandr/tests/kirin_realtime_tests.py:188]: can you add a comment saying that this delay makes the VJ run the day after ?

File: source/kraken/realtime.cpp [hunk 3]
```diff
@@ -198,25 +226,12 @@ create_disruption(const std::string& id,
...
-
-                arrival_seconds +=  get_diff_days(arrival_time) * navitia::DateTimeUtils::SECONDS_PER_DAY;
-                departure_seconds += get_diff_days(departure_time) * navitia::DateTimeUtils::SECONDS_PER_DAY;
+                auto ptime_arrival = bpt::from_time_t(arrival_time) - start_first_day_of_impact;
+                auto ptime_departure = bpt::from_time_t(departure_time) - start_first_day_of_impact;
 
-                type::StopTime stop_time{static_cast<uint32_t>(arrival_seconds),
-                    static_cast<uint32_t>(departure_seconds), stop_point_ptr};
+                type::StopTime stop_time{static_cast<uint32_t>(ptime_arrival.total_seconds()),
+                                         static_cast<uint32_t>(ptime_departure.total_seconds()),
```
Comment (TeXitoi) [inline source/kraken/realtime.cpp]: I prefere using the constructor syntax for that (`uint32_t(ptime_arrival.total_seconds())`) but do as you want.

File: source/kraken/realtime.cpp [hunk 4]
```diff
@@ -43,20 +43,58 @@ www.navitia.io
...
     if (trip_update.trip().schedule_relationship() == transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED
             && trip_update.stop_time_update_size()) {
+        auto start_date = boost::gregorian::from_undelimited_string(trip_update.trip().start_date());
+        auto start_first_day_of_impact = bpt::ptime(start_date, bpt::time_duration(0, 0, 0, 0));
+        for (const auto& st: trip_update.stop_time_update()) {
+            auto ptime_arrival = bpt::from_time_t(st.arrival().time());
+            auto ptime_departure = bpt::from_time_t(st.departure().time());
+            if (ptime_arrival < start_first_day_of_impact
+                    || ptime_departure < start_first_day_of_impact) {
+                LOG4CPLUS_WARN(log, "Stop time " << st.stop_id() << " is before the day of impact");
```
Comment (antoine-de) [inline source/kraken/realtime.cpp]: I think it would be nice the have the id of the trip_update in the log

File: source/kraken/realtime.cpp [hunk 5]
```diff
@@ -43,20 +43,58 @@ www.navitia.io
...
+        }
         return true;
     }
     return false;
 }
 
+static bool check_trip_update(const transit_realtime::TripUpdate& trip_update) {
+    auto log = log4cplus::Logger::getInstance("realtime");
+    if (trip_update.trip().schedule_relationship() ==
+                                transit_realtime::TripDescriptor_ScheduleRelationship_SCHEDULED
```
Comment (antoine-de) [inline source/kraken/realtime.cpp:78]: minor: maybe we could use aliases for those insanely huge enum

File: source/kraken/realtime.cpp [hunk 6]
```diff
@@ -43,20 +43,58 @@ www.navitia.io
...
+            && trip_update.stop_time_update_size()) {
+        uint32_t last_st_dep = std::numeric_limits<uint32_t>::max();
+        for (const auto& st: trip_update.stop_time_update()) {
+            uint32_t arrival_time = st.arrival().time();
+            uint32_t departure_time = st.departure().time();
+            if (last_st_dep != std::numeric_limits<uint32_t>::max()
+                    && last_st_dep > arrival_time) {
+                LOG4CPLUS_WARN(log, "Stop time " << st.stop_id() << " is not correctly ordered");
```
Comment (antoine-de) [inline source/kraken/realtime.cpp]: same here, I think the trip id would be nice in the log

File: source/type/type.cpp [hunk 7]
```diff
@@ -373,10 +374,20 @@ VJ* MetaVehicleJourney::impl_create_vj(const std::string& uri,
...
+    int day_offset = 0;
+    if (!sts.empty()) {
+        day_offset = sts.front().arrival_time / (ndtu::SECONDS_PER_DAY);
+    }
+    ValidityPattern model_new_vp{disrupted_vp};
+    if (day_offset >= 0) {
+        model_new_vp.days <<= size_t(day_offset);
+    } else {
+        model_new_vp.days >>= size_t(-day_offset);
+    }
```
Comment (TeXitoi) [inline source/type/type.cpp:413]: directly using a size_t for day_offset should do the trick as everything is unsigned for the moment. And It won't work as if if arrival_time is signed.

File: source/type/type.cpp [hunk 8]
```diff
@@ -388,10 +399,21 @@ VJ* MetaVehicleJourney::impl_create_vj(const std::string& uri,
         st.set_is_frequency(std::is_same<VJ, FrequencyVehicleJourney>::value);
     }
     vj_ptr->stop_time_list = std::move(sts);
+    // as date management is taken care of in validity pattern,
+    // we have to contain first stop_time in [00:00 ; 24:00[ (and propagate to other st)
+    for (nt::StopTime& st: vj_ptr->stop_time_list) {
+        st.arrival_time -= ndtu::SECONDS_PER_DAY * day_offset;
+        st.departure_time -= ndtu::SECONDS_PER_DAY * day_offset;
+        // a vj cannot be longer than 24h and its start is contained in [00:00 ; 24:00[
```
Comment (TeXitoi) [inline source/type/type.cpp]: in [00:00 ; 48:00[

File: source/kraken/realtime.cpp (hunk 5 above)
Comment (pbougue) [inline source/kraken/realtime.cpp:78]: Yep just dealt with line length, looks bad, but, you know, open source code and all...

**Comment by antoine-de** [comment]
Reviewed 2 of 3 files at r2, 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/canaltp/navitia/1291)_

**Comment by pbougue** [comment]
Jenkins, test this please

File: source/kraken/tests/realtime_test.cpp [hunk 9]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
+
+    navitia::handle_realtime("delay1DayD0", timestamp, trip_update, *b.data);
+
+    // We should have 2 vj
+    BOOST_CHECK_EQUAL(pt_data->vehicle_journeys.size(), 2);
+    BOOST_CHECK_EQUAL(pt_data->routes.size(), 1);
+    BOOST_CHECK_EQUAL(pt_data->lines.size(), 1);
+    // We should have two vp
+    // a vp for the current vj, and an empty vp
+    BOOST_CHECK_GE(pt_data->validity_patterns.size(), 3);
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:746]: unsyncronised comment

File: source/kraken/tests/realtime_test.cpp [hunk 10]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
+
+    navitia::handle_realtime("delay1DayPassMidnightD0", timestamp, trip_update, *b.data);
+
+    // We should have 2 vj
+    BOOST_CHECK_EQUAL(pt_data->vehicle_journeys.size(), 2);
+    BOOST_CHECK_EQUAL(pt_data->routes.size(), 1);
+    BOOST_CHECK_EQUAL(pt_data->lines.size(), 1);
+    // We should have two vp
+    // a vp for the current vj, and an empty vp
+    BOOST_CHECK_GE(pt_data->validity_patterns.size(), 3);
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:813]: again

File: source/kraken/tests/realtime_test.cpp [hunk 11]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
+
+    navitia::handle_realtime("delay1DayD0", timestamp, first_trip_update, *b.data);
+
+    // We should have 2 vj
+    BOOST_CHECK_EQUAL(pt_data->vehicle_journeys.size(), 2);
+    BOOST_CHECK_EQUAL(pt_data->routes.size(), 1);
+    BOOST_CHECK_EQUAL(pt_data->lines.size(), 1);
+    // We should have two vp
+    // a vp for the current vj, and an empty vp
+    BOOST_CHECK_GE(pt_data->validity_patterns.size(), 4);
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:888]: worst

File: source/kraken/tests/realtime_test.cpp [hunk 12]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:968]: bla bla

File: source/kraken/tests/realtime_test.cpp [hunk 13]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:1048]: bobette

File: source/kraken/tests/realtime_test.cpp [hunk 14]
```diff
@@ -710,3 +710,597 @@ BOOST_AUTO_TEST_CASE(invalid_delay) {
...
```
Comment (TeXitoi) [inline source/kraken/tests/realtime_test.cpp:1123]: OK, I stop now

File: source/type/type.h [hunk 15]
```diff
@@ -1042,17 +1045,13 @@ struct MetaVehicleJourney: public Header, HasMessages {
             PT_Data& pt_data, const MetaData& meta, const Route* filtering_route = nullptr);
 
     VehicleJourney*
-    get_vj_at_date(RTLevel level, const boost::gregorian::date& date) const;
-    std::vector<VehicleJourney*>
-    get_vjs_in_period(RTLevel level,
-                      const std::vector<boost::posix_time::time_period>& period,
-                      const MetaData& meta, const Route* filtering_route = nullptr) const;
+    get_vj_circulating_at_date(RTLevel level, const boost::gregorian::date& date) const;
```
Comment (TeXitoi) [inline source/type/type.h]: The name is problematic as it's possible to have 2 adapted vj the same day.

**Comment by TeXitoi** [comment]
OK for me.

**Comment by TeXitoi** [comment]
Jenkins, test this please.

**Pull request merged**